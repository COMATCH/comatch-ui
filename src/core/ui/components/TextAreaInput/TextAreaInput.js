import './TextAreaInput.scss';

import React, { useState } from 'react';

import { CharacterCount } from '../CharacterCount';
import { InputError } from '../InputError';
import { InputLabel } from '../InputLabel';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import noop from 'lodash/noop';

const propTypes = {
    /**
     * Additional class names
     */
    className: PropTypes.string,
    characterCountLabel: PropTypes.node,
    display: PropTypes.oneOf(['block', 'inline']),
    /**
     * SVG icon
     */
    icon: PropTypes.node,
    /**
     * HTML id attribute
     */
    id: PropTypes.string,
    inputError: PropTypes.string,
    label: PropTypes.oneOfType([PropTypes.node, PropTypes.string]),
    maxLength: PropTypes.number,
    name: PropTypes.string.isRequired,
    onChange: PropTypes.func.isRequired,
    onFocus: PropTypes.func,
    onBlur: PropTypes.func,
    placeholder: PropTypes.string,
    htmlPlaceholder: PropTypes.node,
    required: PropTypes.bool,
    rows: PropTypes.number,
    value: PropTypes.string,
    withCharacterCount: PropTypes.bool,
};

const defaultProps = {
    className: null,
    characterCountLabel: null,
    display: 'block',
    icon: null,
    id: null,
    inputError: null,
    label: null,
    maxLength: null,
    onFocus: noop,
    onBlur: noop,
    placeholder: '',
    htmlPlaceholder: null,
    required: false,
    rows: 5,
    value: '',
    withCharacterCount: false,
};

export const TextAreaInput = ({
    className,
    characterCountLabel,
    display,
    htmlPlaceholder,
    icon,
    id,
    inputError,
    label,
    maxLength,
    name,
    onChange,
    onFocus,
    onBlur,
    placeholder,
    required,
    rows,
    value,
    withCharacterCount,
}) => {
    const [currentLength, onCurrentLengthChange] = useState(value.length);
    const classes = classNames('Input', 'TextAreaInput', className, display, {
        'has-error': inputError,
        'with-icon': icon,
        'with-html-placeholder': htmlPlaceholder,
    });
    const showHtmlPlaceholder = htmlPlaceholder && !value;

    const onChangeHandler = (event) => {
        const { value: currentValue = '' } = event.target;

        onCurrentLengthChange(currentValue.length);
        onChange(event);
    };

    return (
        <div id={id} className={classes}>
            {label && <InputLabel text={label} required={required} />}
            <textarea
                className="TextAreaInput__input"
                value={value}
                onChange={onChangeHandler}
                onFocus={onFocus}
                onBlur={onBlur}
                placeholder={placeholder}
                maxLength={maxLength}
                rows={rows}
                name={name}
            />
            {icon}
            {inputError && <InputError text={inputError} />}
            {showHtmlPlaceholder && <div className="html-placeholder">{htmlPlaceholder}</div>}
            {!!withCharacterCount && !!maxLength && (
                <CharacterCount
                    className="TextAreaInput__CharacterCount"
                    currentCount={currentLength}
                    label={characterCountLabel}
                    maxLength={maxLength}
                />
            )}
        </div>
    );
};

TextAreaInput.propTypes = propTypes;

TextAreaInput.defaultProps = defaultProps;
