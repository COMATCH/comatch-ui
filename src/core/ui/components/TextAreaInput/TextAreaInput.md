
## TextAreaInput

From [`TextAreaInput.js`](TextAreaInput.js)

A `<textarea>` input.

prop | type | default | required | description
---- | :----: | :-------: | :--------: | -----------
**className** | `String` | `null` | :x: | Additional class names
**display** | `Enum('block','inline')` | `'block'` | :x: | 
**htmlPlaceholder** | `ReactNode` | `null` | :x: | 
**icon** | `ReactNode` | `null` | :x: | SVG icon
**id** | `String` | `null` | :x: | HTML id attribute
**inputError** | `String` | `null` | :x: | 
**label** | `String` | `null` | :x: | 
**maxLength** | `Number` | `null` | :x: | 
**name** | `String` |  | :white_check_mark: | 
**onBlur** | `Function` | `noop` | :x: | 
**onChange** | `Function` |  | :white_check_mark: | 
**onFocus** | `Function` | `noop` | :x: | 
**placeholder** | `String` | `''` | :x: | 
**required** | `Boolean` | `false` | :x: | 
**rows** | `Number` | `5` | :x: | 
**value** | `String` | `''` | :x: | 



