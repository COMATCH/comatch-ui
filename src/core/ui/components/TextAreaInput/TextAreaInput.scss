@import '../../../styles/variables.scss';

.TextAreaInput {
    display: inline-block;
    margin-bottom: $co-inputs-margin-bottom;

    .InputLabel,
    &.block,
    &__input {
        display: block;
    }

    &__input {
        background-color: $co-color-white;
        border: 1px solid $co-color-mid-gray;
        border-radius: 3px;
        box-shadow: $co-input-box-shadow;
        // sizing & whitespace
        font-size: $co-font-sm;
        padding: 9px;
        width: 100%;
        // Transition effects
        transition: box-shadow 250ms ease-out;
        transform: translateZ(0);
        will-change: box-shadow;
        // reset browser defaults
        outline: 0 !important;
        min-height: 40px;
        resize: vertical;

        &:focus {
            box-shadow: $co-input-box-shadow-focus;
        }
    }

    &__CharacterCount {
        margin-top: 10px;
    }

    // States
    &.has-error {
        .TextAreaInput__input {
            box-shadow: $co-input-box-shadow-error;
        }
    }

    &.with-icon {
        position: relative;

        .TextAreaInput__input {
            padding-right: 30px;
        }

        // icon
        .TextAreaInput__icon,
        svg {
            top: 33px;
            right: 10px;
            position: absolute;
            width: 14px;
            height: 14px;
            pointer-events: none;
            transform: translateY(-23px);
        }

        .InputLabel ~ .TextAreaInput__icon,
        .InputLabel ~ svg {
            top: 58px;
        }
    }

    &.with-html-placeholder {
        position: relative;
        .html-placeholder {
            position: absolute;
            top: 0;
            left: 0;
            padding: 11px 9px;
            font-size: $co-font-sm;
            color: $co-color-mid-gray;
            pointer-events: none;
        }
    }
}

// Defining the color of the placeholder
.TextAreaInput::placeholder {
    /* Chrome, Firefox, Opera, Safari 10.1+ */
    color: $co-color-mid-gray;
    opacity: 1; /* Firefox */
}

.TextAreaInput:-ms-input-placeholder {
    /* Internet Explorer 10-11 */
    color: $co-color-mid-gray;
}

.TextAreaInput::-ms-input-placeholder {
    /* Microsoft Edge */
    color: $co-color-mid-gray;
}
