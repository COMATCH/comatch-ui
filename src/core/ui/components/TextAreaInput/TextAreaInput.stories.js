import React, { Component, useState } from 'react';
import { action } from '@storybook/addon-actions';
import { storiesOf } from '@storybook/react';
import { TextAreaInput } from './TextAreaInput';

const handleChange = (event) => action('changed')(event.target.name, event.target.value);

const html = (
    <span>
        <strong>strong</strong> <br /> text <br /> more text
    </span>
);

storiesOf('TextAreaInput', module)
    .add('with basic style', () => <TextAreaInput name="field" onChange={handleChange} />)
    .add('with placeholder', () => (
        <TextAreaInput name="field" onChange={handleChange} placeholder="Placeholder text" />
    ))
    .add('with html-placeholder', () => <TextAreaInput name="field" onChange={handleChange} htmlPlaceholder={html} />)
    .add('with label', () => <TextAreaInput name="field" onChange={handleChange} label="Label text" />)
    .add('with error', () => (
        <TextAreaInput name="field" onChange={handleChange} label="Label text" inputError="Error text" />
    ))
    .add('with character count', () => <Controlled />)
    .add('in a container', () => <TextAreaInputContainer />);

class TextAreaInputContainer extends Component {
    state = { description: 'this is my text' };

    handleChange = (event) => {
        this.setState({ description: event.target.value });
        action('changed in container')(event.target.name, event.target.value);
    };

    handleFocus = (event) => {
        action('focused in container')(event);
    };

    handleBlur = (event) => {
        action('blured in container')(event);
    };

    render() {
        return (
            <TextAreaInput
                name="description"
                onChange={this.handleChange}
                onFocus={this.handleFocus}
                onBlur={this.handleBlur}
                value={this.state.description}
            />
        );
    }
}

const Controlled = () => {
    const [value, onChange] = useState('');
    const onChangeHandler = (event) => {
        onChange(event.target.value);
        action('changed in `Controlled`')(event.target.name, event.target.value);
    };

    return (
        <TextAreaInput
            withCharacterCount
            characterCountLabel={<span>My custom label:</span>}
            name="field"
            maxLength={2000}
            onChange={onChangeHandler}
            value={value}
        />
    );
};
