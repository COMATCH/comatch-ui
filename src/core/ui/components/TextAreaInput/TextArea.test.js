import React from 'react';
import renderer from 'react-test-renderer';
import noop from 'lodash/noop';
import { mount } from 'enzyme';
import '../../../config/tests/setup';

import { TextAreaInput } from './TextAreaInput';

describe('TextAreaInput', () => {
    const REQUIRED_PROPS = {
        name: 'testing-name',
        onChange: jest.fn(),
    };

    const OPTIONAL_PROPS = {
        className: 'testing-classname',
        display: 'inline',
        icon: <i />,
        id: 'testing-id',
        inputError: 'Testing errror',
        label: 'Testing Label',
        maxLength: 150,
        onFocus: jest.fn(),
        onBlur: jest.fn(),
        placeholder: 'Testing placeholder',
        htmlPlaceholder: <span>Testing HTML Placeholder</span>,
        required: true,
        rows: 16,
        value: 'Testing value',
    };

    it('should render correctly', () => {
        const textAreaInputComponent = renderer.create(<TextAreaInput {...REQUIRED_PROPS} />).toJSON();
        expect(textAreaInputComponent).toMatchSnapshot();
    });

    it('should render with default props', () => {
        const props = {
            ...REQUIRED_PROPS,
        };

        const textAreaInputComponent = mount(<TextAreaInput {...props} />);

        // Required props:
        expect(textAreaInputComponent.prop('name')).toEqual(props.name);
        expect(textAreaInputComponent.prop('onChange')).toEqual(props.onChange);

        // Optional props:
        expect(textAreaInputComponent.prop('className')).toEqual(null);
        expect(textAreaInputComponent.prop('display')).toEqual('block');
        expect(textAreaInputComponent.prop('icon')).toEqual(null);
        expect(textAreaInputComponent.prop('id')).toEqual(null);
        expect(textAreaInputComponent.prop('inputError')).toEqual(null);
        expect(textAreaInputComponent.prop('label')).toEqual(null);
        expect(textAreaInputComponent.prop('maxLength')).toEqual(null);
        expect(textAreaInputComponent.prop('onFocus')).toEqual(noop);
        expect(textAreaInputComponent.prop('onBlur')).toEqual(noop);
        expect(textAreaInputComponent.prop('placeholder')).toEqual('');
        expect(textAreaInputComponent.prop('htmlPlaceholder')).toEqual(null);
        expect(textAreaInputComponent.prop('required')).toEqual(false);
        expect(textAreaInputComponent.prop('rows')).toEqual(5);
        expect(textAreaInputComponent.prop('value')).toEqual('');
    });

    it('should render with passed props', () => {
        const props = {
            ...REQUIRED_PROPS,
            ...OPTIONAL_PROPS,
        };

        const textAreaInputComponent = mount(<TextAreaInput {...props} />);

        // Required props:
        expect(textAreaInputComponent.prop('name')).toEqual(props.name);
        expect(textAreaInputComponent.prop('onChange')).toEqual(props.onChange);

        // Optional props:
        expect(textAreaInputComponent.prop('className')).toEqual(props.className);
        expect(textAreaInputComponent.prop('display')).toEqual(props.display);
        expect(textAreaInputComponent.prop('icon')).toEqual(props.icon);
        expect(textAreaInputComponent.prop('id')).toEqual(props.id);
        expect(textAreaInputComponent.prop('inputError')).toEqual(props.inputError);
        expect(textAreaInputComponent.prop('label')).toEqual(props.label);
        expect(textAreaInputComponent.prop('maxLength')).toEqual(props.maxLength);
        expect(textAreaInputComponent.prop('onFocus')).toEqual(props.onFocus);
        expect(textAreaInputComponent.prop('onBlur')).toEqual(props.onBlur);
        expect(textAreaInputComponent.prop('placeholder')).toEqual(props.placeholder);
        expect(textAreaInputComponent.prop('htmlPlaceholder')).toEqual(props.htmlPlaceholder);
        expect(textAreaInputComponent.prop('required')).toEqual(props.required);
        expect(textAreaInputComponent.prop('rows')).toEqual(props.rows);
        expect(textAreaInputComponent.prop('value')).toEqual(props.value);
    });

    it('should render correctly with a Label', () => {
        const inputWithLabel = (
            <TextAreaInput
                {...REQUIRED_PROPS}
                label={OPTIONAL_PROPS.label}
            />
        );
        const textInputComponentWithLabel = renderer.create(inputWithLabel).toJSON();
        const mountedTextInput = mount(inputWithLabel);
        const textInputContainerNode = mountedTextInput.children().at(0);

        expect(textInputComponentWithLabel).toMatchSnapshot();
        expect(textInputContainerNode.children()).toHaveLength(2);
        expect(textInputContainerNode.children().at(0).name()).toBe('InputLabel');
        expect(textInputContainerNode.children().at(1).name()).toBe('textarea');
    });

    it('should render correctly with an Icon', () => {
        const inputWithIcon = (
            <TextAreaInput
                {...REQUIRED_PROPS}
                icon={OPTIONAL_PROPS.icon}
            />
        );
        const mountedTextInput = mount(inputWithIcon);
        const textInputContainerNode = mountedTextInput.children().at(0);
        const textInputComponentWithIcon = renderer.create(inputWithIcon).toJSON();

        expect(textInputComponentWithIcon).toMatchSnapshot();
        expect(textInputContainerNode.children()).toHaveLength(2);
        expect(textInputContainerNode.children().at(0).name()).toBe('textarea');
        expect(textInputContainerNode.children().at(1).name()).toBe('i');
    });

    it('should render correctly with an InputError', () => {
        const inputWithInputError = (
            <TextAreaInput
                {...REQUIRED_PROPS}
                inputError={OPTIONAL_PROPS.inputError}
            />
        );
        const mountedTextInput = mount(inputWithInputError);
        const textInputContainerNode = mountedTextInput.children().at(0);
        const textInputComponentWithInputError = renderer.create(inputWithInputError).toJSON();

        expect(textInputComponentWithInputError).toMatchSnapshot();
        expect(textInputContainerNode.children()).toHaveLength(2);
        expect(textInputContainerNode.children().at(0).name()).toBe('textarea');
        expect(textInputContainerNode.children().at(1).name()).toBe('InputError');
    });

    it('should render correctly with an HTML Placeholder', () => {
        const inputWitPlaceholder = (
            <TextAreaInput
                {...REQUIRED_PROPS}
                htmlPlaceholder={OPTIONAL_PROPS.htmlPlaceholder}
            />
        );
        const mountedTextInput = mount(inputWitPlaceholder);
        const textInputContainerNode = mountedTextInput.children().at(0);
        const textInputComponentWithInputError = renderer.create(inputWitPlaceholder).toJSON();

        expect(textInputComponentWithInputError).toMatchSnapshot();
        expect(textInputContainerNode.children()).toHaveLength(2);
        expect(textInputContainerNode.children().at(0).name()).toBe('textarea');
        expect(textInputContainerNode.children().at(1).hasClass('html-placeholder')).toBeTruthy();
    });
});
