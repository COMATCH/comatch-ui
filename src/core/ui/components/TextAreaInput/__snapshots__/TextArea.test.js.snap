// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextAreaInput should render correctly 1`] = `
<div
  className="Input TextAreaInput block"
  id={null}
>
  <textarea
    className="TextAreaInput__input"
    maxLength={null}
    name="testing-name"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    placeholder=""
    rows={5}
    value=""
  />
</div>
`;

exports[`TextAreaInput should render correctly with a Character Counter 1`] = `
<div
  className="Input TextAreaInput block"
  id={null}
>
  <textarea
    className="TextAreaInput__input"
    maxLength={2000}
    name="testing-name"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    placeholder=""
    rows={5}
    value=""
  />
  <div
    className="TextAreaInput__CharacterCount"
  >
     
    2000
  </div>
</div>
`;

exports[`TextAreaInput should render correctly with a Label 1`] = `
<div
  className="Input TextAreaInput block"
  id={null}
>
  <label
    className="InputLabel sc-bwzfXH bIVinQ"
    htmlFor={null}
    id={null}
  >
    Testing Label
  </label>
  <textarea
    className="TextAreaInput__input"
    maxLength={null}
    name="testing-name"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    placeholder=""
    rows={5}
    value=""
  />
</div>
`;

exports[`TextAreaInput should render correctly with an HTML Placeholder 1`] = `
<div
  className="Input TextAreaInput block with-html-placeholder"
  id={null}
>
  <textarea
    className="TextAreaInput__input"
    maxLength={null}
    name="testing-name"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    placeholder=""
    rows={5}
    value=""
  />
  <div
    className="html-placeholder"
  >
    <span>
      Testing HTML Placeholder
    </span>
  </div>
</div>
`;

exports[`TextAreaInput should render correctly with an Icon 1`] = `
<div
  className="Input TextAreaInput block with-icon"
  id={null}
>
  <textarea
    className="TextAreaInput__input"
    maxLength={null}
    name="testing-name"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    placeholder=""
    rows={5}
    value=""
  />
  <i />
</div>
`;

exports[`TextAreaInput should render correctly with an InputError 1`] = `
<div
  className="Input TextAreaInput block has-error"
  id={null}
>
  <textarea
    className="TextAreaInput__input"
    maxLength={null}
    name="testing-name"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    placeholder=""
    rows={5}
    value=""
  />
  <span
    className="InputError sc-bdVaJa camiNS"
  >
    Testing errror
  </span>
</div>
`;
